File: p035/p35.c
#include <stdio.h>
#include <memory.h>
#include <stdbool.h>

#include "../lib/prime.h"

void rotate(char* digits, int len) {
	char first = digits[0];
	memmove(&digits[0], &digits[1], len - 1);
	digits[len - 1] = first;
}

bool is_circular_prime(int prime) {
	if(!is_prime(prime)) return false;

	if(prime < 10) return true;

	char digits[64];
	int len = sprintf(digits, "%i", prime);
	rotate(digits, len);

	for(int i = 1; i < len; i++) {
		int n;
		sscanf(digits, "%i", &n);
		if(!is_prime(n)) {
			return false;
		}

		rotate(digits, len);
	}

	return true;
}

bool contains_even_digits(int n) {
	while(n > 0) {
		if(n % 2 == 0) return true;
		n /= 10;
	}
	return false;
}

int main(int argc, char* argv[]) {

	int count = 1;	// count '2' implicitly

	for(int i = 3; i < 1000000; i += 2) {
		if(!contains_even_digits(i) && is_circular_prime(i)) {
			count++;
		}
	}

	printf("%i", count);

	return 0;
}

File: p035/p35.php
<?php

/**
The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 791, 
are themselves prime.

There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, 97.

How many circular primes are there below one million?
**/

require_once "../lib/prime.php";

function is_circular_prime($prime, $circular_primes) {

	if($prime < 10) return true;

	
	$str = strval($prime);
	$str = substr($str, 1) . $str[0];

	for($i = 1; $i < strlen($str); $i++) {
		if(!array_key_exists($str, $circular_primes) &&
		   !is_prime(intval($str))) {
			return false;
		}

		$str = substr($str, 1) . $str[0];
	}	
	return true;
}

function contains_even_digits($n) {
	$str = strval($n);
	$len = strlen($str);
	for($i = 0; $i < $len; $i++) {
		if($str[$i] % 2 == 0) return true;
	}
	return false;
}

// first find all primes below 1 million
$circular_primes = array(2);
for($i = 3; $i < 1000000; $i += 2) {
	if(!contains_even_digits($i) && is_prime($i) && is_circular_prime($i, $circular_primes)) {
		$circular_primes[$i] = $i;
	}	
}

print count($circular_primes);

?>
